### PR TITLE
fix(connection-limits): correct connection tracking

### LINKED ISSUE #4249: Failed outbound connections and connection limits
## Summary

This may be a pre-mature issue as I've not yet found the actual source of the issue. 

Lighthouse is upgrading to libp2p v0.52. We cannot establish new connections. What appears to be happening is:
1. We attempt to dial a multitude of peers. 
2. Many of these fail (they can be offline, or not accepting connections. We often see transport failures (I believe this is normal). For example: ` Failed to negotiate transport protocol(s): [(/ip4/x.x.x.x/tcp/9000: : Multistream select failed: Multistream select failed: Protocol error: I/O error: Connection reset by peer (os error 104): I/O error: Connection reset by peer (os error 104): Connection reset by peer (os error 104))]`
3.  Lighthouse has an outbound pending connection limit of 16. After 16 failed dial attempts, all future attempts fail with:
`Denied { cause: ConnectionDenied { inner: Exceeded { limit: 16, kind: PendingOutgoing } } }`

Initial suspicion is that these failures are not reducing the outbound connection count. I'm posting this issue before I check the latest code, as someone may see the problem faster than me.

### PR DESCRIPTION
**Pull request by AgeManning**
## Description

The connection limit behaviour was not taking into account connection errors. Also, it was using the behaviour events to indicate established connections which is not always going to be the case because other behaviours can deny the connections (thanks @divagant-martian). 

Closes #4249

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: misc/connection-limits/CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,12 @@
-## 0.2.0 
+## 0.2.0
```
Comment (dariusc93) [inline misc/connection-limits/CHANGELOG.md:14]: Shouldnt this be `0.2.1` (unreleased)?

Comment (AgeManning) [reply misc/connection-limits/CHANGELOG.md:14]: Yeah, good point.

File: misc/connection-limits/CHANGELOG.md [hunk 2]
```diff
@@ -1,12 +1,16 @@
-## 0.2.0
+## 0.2.1 (unreleased)
```
Comment (dariusc93) [inline misc/connection-limits/CHANGELOG.md]:
```suggestion
## 0.2.1 - unreleased
```

**Comment by mxinden** [comment]
Thank you @AgeManning. Reviewing right now. Will prioritize a patch release.

**Comment by mxinden** [comment]
//CC @shamil-gadelshin as I think you have described seeing this bug in the last open maintainers call.

File: misc/connection-limits/src/lib.rs [hunk 3]
```diff
@@ -327,11 +325,43 @@ impl NetworkBehaviour for Behaviour {
                     .entry(peer_id)
                     .or_default()
                     .remove(&connection_id);
+
+                self.pending_inbound_connections.remove(&connection_id);
+                self.pending_outbound_connections.remove(&connection_id);
```
Comment (mxinden) [inline misc/connection-limits/src/lib.rs]: @AgeManning are these two lines needed? A `FromSwarm::ConnectionClosed` should always be preceded by a `handle_established_{outbound,inbound}_connection`. In those methods we reduce the `self.pending_*_connections` counters.

Comment (mxinden) [reply misc/connection-limits/src/lib.rs]: Removed with 8167c711e. Please still review @AgeManning.

**Comment by mxinden** [comment]
I pushed a couple of commits. Hope you don't mind @AgeManning. Any objections to the additional changes? If not, I will merge here and cut a patch release.

Again, thank you for the work @AgeManning and @divagant-martian.

**Review by divagant-martian — APPROVED** [review]
Ty for the follow up changes @mxinden. To speed this up since it's late for @AgeManning I'm taking the lead on saying this looks good

File: misc/connection-limits/src/lib.rs (hunk 3 above)
Comment (divagant-martian) [reply misc/connection-limits/src/lib.rs]: I agree this is fine to remove. Just being overly cautious

File: misc/connection-limits/CHANGELOG.md [hunk 4]
```diff
@@ -1,4 +1,18 @@
-## 0.2.0 
+## 0.2.1 - unreleased
```
Comment (mxinden) [inline misc/connection-limits/CHANGELOG.md]:
```suggestion
## 0.2.1
```

Adding ` - unreleased` is the way to go. Thank you. I will cut a release right away, thus removing it already.

File: misc/connection-limits/src/lib.rs [hunk 5]
```diff
@@ -466,19 +480,126 @@ mod tests {
...
+            cause.downcast::<std::io::Error>().unwrap();
+
+            assert_eq!(
+                0,
+                swarm1
+                    .behaviour_mut()
+                    .limits
+                    .established_inbound_connections
+                    .len()
+                    ,
```
Comment (mxinden) [inline misc/connection-limits/src/lib.rs]:
```suggestion
                    .len(),
```

**Review by mxinden — DISMISSED** [review]

**Review by mxinden — DISMISSED** [review]

**Comment by divagant-martian** [comment]
@mxinden just to bring your attention to the fact that mergify was not very fond of the merge commit

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
> @mxinden just to bring your attention to mergify was not very fond of the merge commit

:/ Thanks! I think the root of the issue is, that only I have push access to https://github.com/sigp/, not all rust-libp2p maintainers (including mergify).

**Comment by mxinden** [comment]
[Tagged](https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-connection-limits-v0.2.1) and [published](https://crates.io/crates/libp2p-connection-limits/0.2.1).

Thank you for the debugging and implementation work.

**Comment by AgeManning** [comment]
Nice. Yeah I was sleeping, sorry. Looks like diva had us covered.

**Review by thomaseizinger — COMMENTED** [review]
Thank you for the effort in debugging and fixing this!

**Pull request merged**